### PR TITLE
feat(rules): add require-top-level-describe rule

### LINKED ISSUE #401: [new-rule] Enforce top level describe
Hello everyone,

Jest, and most testing library, let you write and organise your tests as you want, which is good but it can become a mess with large testing codebase.
(Somewhat similar to #362 and `no-standalone-expect` ) 
It would be super cool to have a rule to enforce a top-level describe and make sure no `test`, `it`, `beforeAll` etc... can live outside this describe.
Maybe it can be something like:
- `require-top-level-describe` 
or 
- merged with `no-standalone-expect` 
   => `no-standalone: ['error', ['test', 'hook', 'expect']]`

What do you think?

### PR DESCRIPTION
**Pull request by thomlom**
This PR intends to add a new ESLint rule which enforces the user to add a top-level `describe` in their test files.

Closes #401

### CODE REVIEW

**Comment by SimenB** [comment]
Thanks! Ignore CI, the commitlint thing is somewhat broken. I should look into it at some point...

File: src/rules/require-top-level-describe.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    },
+    type: 'suggestion',
+    schema: [],
+  },
+  defaultOptions: [],
+  create(context) {
+    let numberOfDescribeBlock = 0;
+    return {
+      CallExpression(node) {
+        isDescribe(node) && numberOfDescribeBlock++;
```
Comment (SimenB) [inline src/rules/require-top-level-describe.ts]:
```suggestion
        if (isDescribe(node)) {
          numberOfDescribeBlock++;

          return;
        }
```

File: src/rules/require-top-level-describe.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  name: __filename,
+  meta: {
+    docs: {
+      category: 'Best Practices',
+      description:
+        'Prevents test cases and hooks to be outside of a describe block',
+      recommended: false,
+    },
+    messages: {
+      unexpectedMethod: 'There must be a top-level describe block.',
```
Comment (SimenB) [inline src/rules/require-top-level-describe.ts]:
```suggestion
      unexpectedMethod: 'All test cases must be wrapped in a describe block.',
```
?

File: src/rules/require-top-level-describe.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    type: 'suggestion',
+    schema: [],
+  },
+  defaultOptions: [],
+  create(context) {
+    let numberOfDescribeBlock = 0;
+    return {
+      CallExpression(node) {
+        isDescribe(node) && numberOfDescribeBlock++;
+        if ((isTestCase(node) || isHook(node)) && numberOfDescribeBlock === 0) {
```
Comment (SimenB) [inline src/rules/require-top-level-describe.ts]:
```suggestion
        if (numberOfDescribeBlock === 0 && (isTestCase(node) || isHook(node))) {
```

cheapest check, so might as well bail early if possible

File: src/rules/require-top-level-describe.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    },
+    messages: {
+      unexpectedMethod: 'There must be a top-level describe block.',
+    },
+    type: 'suggestion',
+    schema: [],
+  },
+  defaultOptions: [],
+  create(context) {
+    let numberOfDescribeBlock = 0;
```
Comment (SimenB) [inline src/rules/require-top-level-describe.ts]:
```suggestion
    let numberOfDescribeBlocks = 0;
```

File: src/rules/require-top-level-describe.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  name: __filename,
+  meta: {
+    docs: {
+      category: 'Best Practices',
+      description:
+        'Prevents test cases and hooks to be outside of a describe block',
+      recommended: false,
+    },
+    messages: {
+      unexpectedMethod: 'All test cases must be wrapped in a describe block.',
```
Comment (SimenB) [inline src/rules/require-top-level-describe.ts]: should we have separate test case and hook error message?

File: src/rules/require-top-level-describe.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    let numberOfDescribeBlocks = 0;
+    return {
+      CallExpression(node) {
+        if (isDescribe(node)) {
+          numberOfDescribeBlocks++;
+          return;
+        }
+
+        if (
+          (numberOfDescribeBlocks === 0 && isTestCase(node)) ||
```
Comment (SimenB) [inline src/rules/require-top-level-describe.ts]: this is wrong, right?
```suggestion
          (numberOfDescribeBlocks === 0) ||
           (isTestCase(node) || sHook(node))
```
looks more correct

File: src/rules/require-top-level-describe.ts (hunk 5 above)
Comment (thomlom) [reply src/rules/require-top-level-describe.ts]: We could go with `'All test cases and hooks must be wrapped in a describe block.'` but it feels long to me.
Having separate error messages make sense to me. I'll add a test in that case!

File: src/rules/require-top-level-describe.ts (hunk 6 above)
Comment (thomlom) [reply src/rules/require-top-level-describe.ts]: Oops, my mistake 🙂

File: src/rules/require-top-level-describe.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+          return;
+        }
+
+        if (numberOfDescribeBlocks === 0) {
+          if (isTestCase(node)) {
+            context.report({ node, messageId: 'unexpectedTestCase' });
+            return;
+          }
+
+          if (isHook(node)) {
```
Comment (SimenB) [inline src/rules/require-top-level-describe.ts]: CI complains that this is not covered. Maybe the `else` branch? A test with a `CallExpression` inside a `describe` that's neither a test nor a hook should probably hit it?

Comment (thomlom) [reply src/rules/require-top-level-describe.ts]: Indeed! I added a dummy test to fix it: 

```js
'foo()',
```

But thinking of it, should we trigger a warning if the user uses a `CallExpression` that's neither a test or a hook or can we leave it as is?

Comment (SimenB) [reply src/rules/require-top-level-describe.ts]: no, that's fine, I think. Jest itself uses it to e.g. skip tests on windows, or run some other sort of helper 🙂

**Pull request merged**